### PR TITLE
Validate the category in OSSL_trace_end()

### LINKED ISSUE #17353: Potential locking problems due to the inconsistent checking
Hi, I found a potential locking problem.

When `category < 0` at line 466, the lock `trace_lock`at Line 473 is not acquired.
https://github.com/openssl/openssl/blob/a935791d54078f43209ffbc1886ac5e68772ce34/crypto/trace.c#L459-L473

The range of `category` is not checked at line 498, lock `trace_lock` could be released without acquiring at line 516?
https://github.com/openssl/openssl/blob/a935791d54078f43209ffbc1886ac5e68772ce34/crypto/trace.c#L493-L516

### PR DESCRIPTION
**Pull request by mattcaswell**
OSSL_trace_end() should validate that the category it has been passed
by the caler is valid, and return immediately if not.

Fixes #17353

### CODE REVIEW

**Review by t8m — APPROVED** [review]

File: crypto/trace.c [hunk 1]
```diff
@@ -496,6 +496,8 @@ void OSSL_trace_end(int category, BIO * channel)
     char *suffix = NULL;
 
     category = ossl_trace_get_category(category);
+    if (category < 0)
+        return;
```
Comment (mspncp) [inline crypto/trace.c:500]: It‘s a programming error if the trace groups are not properly nested. Instead of silently ignoring it, I‘d add `ossl_assert(category >= 0)`. (Note that there is already an assertion a few lines below). 

(Disclaimer: this is a quick drive-by comment from m mobile phone)

**Review by mspncp — CHANGES_REQUESTED** [review]

File: crypto/trace.c (hunk 1 above)
Comment (mspncp) [reply crypto/trace.c:500]: Hmmm… just read issue #17353. I need to take another look when I‘m back on my computer.

Comment (mattcaswell) [reply crypto/trace.c:500]: We normally reserve `ossl_assert` and `assert` for catching internal errors. This is actually public API, and so the "bug" would be in the caller and therefore may not be an "internal" error. So, its not clear to me that `ossl_assert` is actually appropriate.

**Review by mspncp — APPROVED** [review]
Ok, never mind my first change request. One could turn it into an assertion (both in OSSL_trace_begin and OSSL_trace_end), but it's ok as as it is as well.

File: crypto/trace.c (hunk 1 above)
Comment (t-j-h) [reply crypto/trace.c:500]: It is not an *assert* context. This is something that can easily happen with a public API.

**Review by t-j-h — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master and 3.0.
